README.md
## Features

Record you coding time.

## How to use

After install this extension, you can find `Show VSRecord` in your editor context menu:

<p><img src="https://github.com/leasontou/vsrecoder/raw/master/images/menu.jpg" alt="menu" width="280px"></p>

## Screenshot

Your coding heatmap:

<p><img src="https://github.com/leasontou/vsrecoder/raw/master/images/record.jpg" alt="record" width="600px"></p>

Check your contributions in each durations:

<p><img src="https://github.com/leasontou/vsrecoder/raw/master/images/tips.jpg" alt="tips" width="600px"></p>

Contact me on [GitHub](https://github.com/leasontou/vsrecoder)


CHANGELOG.md
# Change Log

All notable changes to the "vsrecode" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/util.js
const fs = require('fs');
const os = require('os');
const path = require('path');
const vscode = require('vscode');
const exec = require('child_process').exec;

const util = {
    dateFormat(fmt, date) {
        let ret;
        let opt = {
            "Y+": date.getFullYear().toString(),        // 年
            "m+": (date.getMonth() + 1).toString(),     // 月
            "d+": date.getDate().toString(),            // 日
            "H+": date.getHours().toString(),           // 时
            "M+": date.getMinutes().toString(),         // 分
            "S+": date.getSeconds().toString()          // 秒
            // 有其他格式化字符需求可以继续添加，必须转化成字符串
        };
        for (let k in opt) {
            ret = new RegExp("(" + k + ")").exec(fmt);
            if (ret) {
                fmt = fmt.replace(ret[1], (ret[1].length == 1) ? (opt[k]) : (opt[k].padStart(ret[1].length, "0")))
            };
        };
        return fmt;
    },
    /**
     * 获取当前所在工程根目录，有3种使用方法：<br>
     * getProjectPath(uri) uri 表示工程内某个文件的路径<br>
     * getProjectPath(document) document 表示当前被打开的文件document对象<br>
     * getProjectPath() 会自动从 activeTextEditor 拿document对象，如果没有拿到则报错
     * @param {*} document 
     */
    getProjectPath(document) {
        if (!document) {
            document = vscode.window.activeTextEditor ? vscode.window.activeTextEditor.document : null;
        }
        if (!document) {
            this.showError('当前激活的编辑器不是文件或者没有文件被打开！');
            return '';
        }
        const currentFile = (document.uri ? document.uri : document).fsPath;
        let projectPath = null;

        let workspaceFolders = vscode.workspace.workspaceFolders.map(item => item.uri.path);
        // 由于存在Multi-root工作区，暂时没有特别好的判断方法，先这样粗暴判断
        // 如果发现只有一个根文件夹，读取其子文件夹作为 workspaceFolders
        if (workspaceFolders.length == 1 && workspaceFolders[0] === vscode.workspace.rootPath) {
            const rootPath = workspaceFolders[0];
            var files = fs.readdirSync(rootPath);
            workspaceFolders = files.filter(name => !/^\./g.test(name)).map(name => path.resolve(rootPath, name));
            // vscode.workspace.rootPath会不准确，且已过时
            // return vscode.workspace.rootPath + '/' + this._getProjectName(vscode, document);
        }
        workspaceFolders.forEach(folder => {
            if (currentFile.indexOf(folder) === 0) {
                projectPath = folder;
            }
        })
        if (!projectPath) {
            this.showError('获取工程根路径异常！');
            return '';
        }
        return projectPath;
    },
    /**
     * 获取当前工程名
     */
    getProjectName: function(projectPath) {
        return path.basename(projectPath);
    },
    getPluginPath() {

    },
    /**
     * 将一个单词首字母大写并返回
     * @param {*} word 某个字符串
     */
    upperFirstLetter: function(word) {
        return (word || '').replace(/^\w/, m => m.toUpperCase());
    },
    /**
     * 将一个单词首字母转小写并返回
     * @param {*} word 某个字符串
     */
    lowerFirstLeter: function(word) {
        return (word || '').replace(/^\w/, m => m.toLowerCase());
    },
    /**
     * 全局日志开关，发布时可以注释掉日志输出
     */
    log: function(...args) {
        console.log(...args);
    },
    /**
     * 全局日志开关，发布时可以注释掉日志输出
     */
    error: function(...args) {
        console.error(...args);
    },
    /**
     * 弹出错误信息
     */
    showError: function(info) {
        vscode.window.showErrorMessage(info);
    },
    /**
     * 弹出提示信息
     */
    showInfo: function(info) {
        vscode.window.showInformationMessage(info);
    },
    findStrInFolder: function(folderPath, str) {

    },
    /**
     * 从某个文件里面查找某个字符串，返回第一个匹配处的行与列，未找到返回第一行第一列
     * @param filePath 要查找的文件
     * @param reg 正则对象，最好不要带g，也可以是字符串
     */
    findStrInFile: function(filePath, reg) {
        const content = fs.readFileSync(filePath, 'utf-8');
        reg = typeof reg === 'string' ? new RegExp(reg, 'm') : reg;
        // 没找到直接返回
        if (content.search(reg) < 0) return {row: 0, col: 0};
        const rows = content.split(os.EOL);
        // 分行查找只为了拿到行
        for(let i = 0; i < rows.length; i++) {
            let col = rows[i].search(reg);
            if(col >= 0) {
                return {row: i, col};
            }
        }
        return {row: 0, col: 0};
    },
    /**
     * 获取某个字符串在文件里第一次出现位置的范围，
     */
    getStrRangeInFile: function(filePath, str) {
        var pos = this.findStrInFile(filePath, str);
        return new vscode.Range(new vscode.Position(pos.row, pos.col), new vscode.Position(pos.row, pos.col + str.length));
    },
    /**
     * 简单的检测版本大小
     */
    checkVersion: function(version1, version2) {
        version1 = parseInt(version1.replace(/\./g, ''));
        version2 = parseInt(version2.replace(/\./g, ''));
        return version1 > version2;
    },
    /**
     * 获取某个扩展文件绝对路径
     * @param context 上下文
     * @param relativePath 扩展中某个文件相对于根目录的路径，如 images/test.jpg
     */
    getExtensionFileAbsolutePath: function(context, relativePath) {
        return path.join(context.extensionPath, relativePath);
    },
    /**
     * 获取某个扩展文件相对于webview需要的一种特殊路径格式
     * 形如：vscode-resource:/Users/toonces/projects/vscode-cat-coding/media/cat.gif
     * @param context 上下文
     * @param relativePath 扩展中某个文件相对于根目录的路径，如 images/test.jpg
     */
    getExtensionFileVscodeResource: function(context, relativePath) {
        const diskPath = vscode.Uri.file(path.join(context.extensionPath, relativePath));
        return diskPath.with({ scheme: 'vscode-resource' }).toString();
    },
    /**
     * 在Finder中打开某个文件或者路径
     */
    openFileInFinder: function(filePath) {
        if (!fs.existsSync(filePath)) {
            this.showError('文件不存在：' + filePath);
        }
        // 如果是目录，直接打开就好
        if (fs.statSync(filePath).isDirectory()) {
            exec(`open ${filePath}`);
        } else {
            // 如果是文件，要分开处理
            const fileName = path.basename(filePath);
            filePath = path.dirname(filePath);
            // 这里有待完善，还不知道如何finder中如何选中文件
            exec(`open ${filePath}`);
        }
    },
    /**
     * 在vscode中打开某个文件
     * @param {*} path 文件绝对路径
     * @param {*} text 可选，如果不为空，则选中第一处匹配的对应文字
     */
    openFileInVscode: function(path, text) {
        let options = undefined;
        if (text) {
            const selection = this.getStrRangeInFile(path, text);
            options = { selection };
        }
        vscode.window.showTextDocument(vscode.Uri.file(path), options);
    },
    /**
     * 用JD-GUI打开jar包
     */
    openJarByJdGui: function(jarPath) {
        // 如何选中文件有待完善
        const jdGuiPath = vscode.workspace.getConfiguration().get('eggHelper.jdGuiPath');
        if (!jdGuiPath) {
            this.showError('JD-GUI路径不能为空！');
            return;
        }
        if (!fs.existsSync(jdGuiPath)) {
            this.showError('您还没有安装JD-GUI，请安装完后到vscode设置里面找到HSF助手并进行路径配置。');
            return;
        }
        if (!fs.existsSync(jarPath)) {
            this.showError('jar包未找到：' + jarPath);
            return;
        }
        exec(`open ${jarPath} -a ${jdGuiPath}`);
    },
    /**
     * 使用默认浏览器中打开某个URL
     */
    openUrlInBrowser: function(url) {
        exec(`open '${url}'`);
    },
    /**
     * 递归遍历清空某个资源的require缓存
     * @param {*} absolutePath
     */
    clearRequireCache(absolutePath) {
        const root = require.cache[absolutePath];
        if (!root) return;
        if (root.children) {
            // 如果有子依赖项，先清空依赖项的缓存
            root.children.forEach(item => {
                this.clearRequireCache(item.id);
            });
        }
        delete require.cache[absolutePath];
    },
    /**
     * 动态require，和普通require不同的是，加载之前会先尝试删除缓存
     * @param {*} modulePath 
     */
    dynamicRequire(modulePath) {
        this.clearRequireCache(modulePath);
        return require(modulePath);
    },
    /**
     * 读取properties文件
     * @param {*} filePath 
     */
    readProperties(filePath) {
        const content =  fs.readFileSync(filePath, 'utf-8');
        let rows = content.split(os.EOL);
        rows = rows.filter(row => row && row.trim() && !/^#/.test(row));
        const result = {};
        rows.forEach(row => {
            let temp = row.split('=');
            result[temp[0].trim()] = temp[1].trim();
        });
        return result;
    },
    /**
     * 比较2个对象转JSON字符串后是否完全一样
     * 特别注意，由于JS遍历对象的无序性（部分浏览器是按照添加顺序遍历的）导致同样的对象，
     * 转成JSON串之后反而不一样，所以这里采用其它方式实现。
     * @param {*} obj1 
     * @param {*} obj2 
     */
    jsonEquals(obj1, obj2) {
        let s1 = this.formatToSpecialJSON(obj1, '', true);
        let s2 = this.formatToSpecialJSON(obj2, '', true);
        return s1 === s2;
    }
};

module.exports = util;

src/extension.js
const vscode = require('vscode');

const util = require('./util');
const fs = require('fs');
const path = require('path');

const Status = {
	idle: 'idle',
	coding: 'coding',
	debug: 'debug'
}

const vsrecoder = {
	status: Status.idle,
	idleTimer: null,
	codingTimer: null,
	debugTimer: null,
	debugTracker: null,
	isCoding: false,
	isIdle: false,
	cleanRecord(context){
		let now = new Date()
		let day = util.dateFormat('YYYYmmdd',now)
		context.globalState.update(day, undefined)
	},
	getRecord(context,day){
		let data = context.globalState.get(day,[])
		return data
	},
	update(context,status){
		try {
			let now = new Date()
			let day = util.dateFormat('YYYYmmdd',now)
			let oldData = context.globalState.get(day,[])
			if(status==Status.idle && oldData.length>0 && oldData[oldData.length-1].type==Status.idle){
				return
			}
			oldData.push({
				type: status,
				time: now.getTime()
			})
			context.globalState.update(day, oldData)
		} catch (error) {
			util.showError(error)
		}
	},
	startDebug(context){
		this.isIdle = false
		if(this.status!=Status.debug){
			this.status = Status.debug
			if(this.debugTimer){
				clearInterval(this.debugTimer)
			}
			let debugInterval = vscode.workspace.getConfiguration().get("debugInterval",2*60)
			this.debugTimer = setInterval(() => {
				if(this.status==Status.debug){
					this.stopDebug(context)
				}
			}, debugInterval*1000);

			console.log("record debug")
			this.update(context,Status.debug)
		}
	},
	stopDebug(context){
		this.status === Status.idle
		this.isIdle = true
		if(this.debugTimer){
			clearInterval(this.debugTimer)
			this.debugTimer = null
		}
		if(this.debugTracker){
			this.debugTracker.dispose()
		}
	},
	startCoding(context){
		this.isCoding = true
		this.isIdle = false
		if(this.status!=Status.coding){
			this.status = Status.coding
			this.update(context,Status.coding)
			console.log(`record coding`)
			if(this.codingTimer){
				clearInterval(this.codingTimer)
			}
			let codingInterval = vscode.workspace.getConfiguration().get("codingInterval",1*60)
			this.codingTimer = setInterval(() => {
				if(this.status==Status.coding){
					if(this.isCoding){
						this.stopCoding(context)
					}
					this.isIdle = true
					this.isCoding = false
					this.status = Status.idle
				}
			}, codingInterval*1000);
		}
	},
	stopCoding(context){
		this.isCoding = false
		this.status = Status.idle
		if(this.codingTimer){
			clearInterval(this.codingTimer)
			this.codingTimer = null
		}
	},
	startIdle(context){
		this.status = Status.idle
		this.isIdle = true
		if(this.idleTimer){
			clearInterval(this.idleTimer)
		}

		let idleInterval = vscode.workspace.getConfiguration().get("idleInterval",30)
		this.idleTimer = setInterval(() => {
			let now = new Date()
			// console.log(`onIdle:${util.dateFormat('HH:MM:SS',now)},${this.status},isIdle:${this.isIdle},isCoding:${this.isCoding}`)
			if(this.status === Status.idle && this.isIdle){
				console.log(`record idle`)
				this.isIdle = false
				this.update(context,Status.idle)
			}
		}, idleInterval*1000);
	},
	stopIdleTimer(context){
		if(this.idleTimer){
			clearInterval(this.idleTimer)
			this.idleTimer = null
		}
	},
	deactivate(){
		this.stopCoding()
		this.stopIdleTimer()
		this.stopDebug()
	}
}
function getWebViewContent(context, templatePath) {
	const resourcePath = util.getExtensionFileAbsolutePath(context, templatePath);
	const dirPath = path.dirname(resourcePath);
	let html = fs.readFileSync(resourcePath, 'utf-8');
	html = html.replace(/(<link.+?href="|<script.+?src="|<img.+?src=")(.+?)"/g, (m, $1, $2) => {
			return $1 + vscode.Uri.file(path.resolve(dirPath, $2)).with({ scheme: 'vscode-resource' }).toString() + '"';
	});
	return html;
}

function invokeCallback(panel, message, resp) {
	console.log(`${message.cmd}`);
	if (typeof resp == 'object' && resp.code && resp.code >= 400 && resp.code < 600) {
			util.showError(resp.message || 'unknown error');
	}
	panel.webview.postMessage({cmd: 'vscodeCallback', cbid: message.cbid, data: resp});
}

const messageHandler = {
	getConfig(context,global, message) {
		const result = vscode.workspace.getConfiguration().get(message.key);
		invokeCallback(global.panel, message, result);
	},
	setConfig(context,global, message) {
		vscode.workspace.getConfiguration().update(message.key, message.value, true);
		util.showInfo('Setting updated');
	},

	getRecord(context,global, message){
		let day = message.day
		let result = vsrecoder.getRecord(context,day)
		invokeCallback(global.panel, message, result);
	}
};

function activate(context) {
	vsrecoder.startIdle(context)
	vscode.workspace.onDidChangeTextDocument(event => {
		vsrecoder.startCoding(context)
	})
	vscode.debug.onDidStartDebugSession(event => {
		vsrecoder.startDebug(context)
	})
	vscode.debug.onDidTerminateDebugSession(event => {
		vsrecoder.stopDebug(context)
	})
	vsrecoder.debugTracker = vscode.debug.registerDebugAdapterTrackerFactory('*',{
		createDebugAdapterTracker(session,executable){
			return {
				onDidSendMessage(msg){
				},
				onWillReceiveMessage(msg){
					vsrecoder.startDebug(context)
				},
				onWillStartSession(){
				},
				onWillStopSession(){
				}
			}
		}
	})

	context.subscriptions.push(vscode.commands.registerCommand('extension.vsrecoder.showRecord', function (){
		const panel = vscode.window.createWebviewPanel(
				'VSRecorder',
				"VSRecorder",
				vscode.ViewColumn.One,
				{
					enableScripts: true,
				}
		);

		let global = { panel};

		panel.webview.html = getWebViewContent(context, 'src/view/vsrecoder.html');
		panel.webview.onDidReceiveMessage(message => {
				if (messageHandler[message.cmd]) {
						messageHandler[message.cmd](context,global, message);
				} else {
						util.showError(`${message.cmd} not found`);
				}
		}, undefined, context.subscriptions)
	}))
}
exports.activate = activate;

function deactivate() {
	vsrecoder.deactivate()
}

module.exports = {
	activate,
	deactivate
}


js/app.16e3e59a.js
(function(e){function t(t){for(var r,i,u=t[0],c=t[1],l=t[2],p=0,f=[];p<u.length;p++)i=u[p],Object.prototype.hasOwnProperty.call(o,i)&&o[i]&&f.push(o[i][0]),o[i]=0;for(r in c)Object.prototype.hasOwnProperty.call(c,r)&&(e[r]=c[r]);s&&s(t);while(f.length)f.shift()();return a.push.apply(a,l||[]),n()}function n(){for(var e,t=0;t<a.length;t++){for(var n=a[t],r=!0,u=1;u<n.length;u++){var c=n[u];0!==o[c]&&(r=!1)}r&&(a.splice(t--,1),e=i(i.s=n[0]))}return e}var r={},o={app:0},a=[];function i(t){if(r[t])return r[t].exports;var n=r[t]={i:t,l:!1,exports:{}};return e[t].call(n.exports,n,n.exports,i),n.l=!0,n.exports}i.m=e,i.c=r,i.d=function(e,t,n){i.o(e,t)||Object.defineProperty(e,t,{enumerable:!0,get:n})},i.r=function(e){"undefined"!==typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(e,Symbol.toStringTag,{value:"Module"}),Object.defineProperty(e,"__esModule",{value:!0})},i.t=function(e,t){if(1&t&&(e=i(e)),8&t)return e;if(4&t&&"object"===typeof e&&e&&e.__esModule)return e;var n=Object.create(null);if(i.r(n),Object.defineProperty(n,"default",{enumerable:!0,value:e}),2&t&&"string"!=typeof e)for(var r in e)i.d(n,r,function(t){return e[t]}.bind(null,r));return n},i.n=function(e){var t=e&&e.__esModule?function(){return e["default"]}:function(){return e};return i.d(t,"a",t),t},i.o=function(e,t){return Object.prototype.hasOwnProperty.call(e,t)},i.p="/";var u=window["webpackJsonp"]=window["webpackJsonp"]||[],c=u.push.bind(u);u.push=t,u=u.slice();for(var l=0;l<u.length;l++)t(u[l]);var s=c;a.push([0,"chunk-vendors"]),n()})({0:function(e,t,n){e.exports=n("56d7")},2395:function(e,t,n){},"56d7":function(e,t,n){"use strict";n.r(t);n("e260"),n("e6cf"),n("cca6"),n("a79d");var r=n("2b0e"),o=function(){var e=this,t=e.$createElement,n=e._self._c||t;return n("div",{attrs:{id:"app"}},[n("router-view")],1)},a=[],i=(n("7c55"),n("2877")),u={},c=Object(i["a"])(u,o,a,!1,null,null,null),l=c.exports,s=n("8c4f"),p=function(){var e=this,t=e.$createElement,n=e._self._c||t;return n("div",{staticClass:"home"},[n("v-chart",{attrs:{options:e.polar}})],1)},f=[],d={name:"home",data:function(){for(var e=[],t=0;t<=360;t++){var n=t/180*Math.PI,r=Math.sin(2*n)*Math.cos(2*n);e.push([r,t])}var o={title:{text:"极坐标双数值轴"},legend:{data:["line"]},polar:{center:["50%","54%"]},tooltip:{trigger:"axis",axisPointer:{type:"cross"}},angleAxis:{type:"value",startAngle:0},radiusAxis:{min:0},series:[{coordinateSystem:"polar",name:"line",type:"line",showSymbol:!1,data:e}],animationDuration:2e3};return{polar:o}}},v=d,h=(n("6cce"),Object(i["a"])(v,p,f,!1,null,"44b7a02b",null)),b=h.exports;r["a"].use(s["a"]);var m=[{path:"/",name:"home",component:b}],y=new s["a"]({routes:m}),g=y,w=n("2f62");r["a"].use(w["a"]);var x=new w["a"].Store({state:{},mutations:{},actions:{},modules:{}}),O=n("9ca8");n("ef97"),n("2f73");r["a"].config.productionTip=!1,r["a"].component("v-chart",O["a"]),new r["a"]({router:g,store:x,render:function(e){return e(l)}}).$mount("#app")},"6a0b":function(e,t,n){},"6cce":function(e,t,n){"use strict";var r=n("6a0b"),o=n.n(r);o.a},"7c55":function(e,t,n){"use strict";var r=n("2395"),o=n.n(r);o.a}});
//# sourceMappingURL=app.16e3e59a.js.map

src/view/vsrecoder.js

const vscode = acquireVsCodeApi();
const callbacks = {};

function dateFormat(fmt, date) {
    let ret;
    let opt = {
        "Y+": date.getFullYear().toString(),        // 年
        "m+": (date.getMonth() + 1).toString(),     // 月
        "d+": date.getDate().toString(),            // 日
        "H+": date.getHours().toString(),           // 时
        "M+": date.getMinutes().toString(),         // 分
        "S+": date.getSeconds().toString()          // 秒
        // 有其他格式化字符需求可以继续添加，必须转化成字符串
    };
    for (let k in opt) {
        ret = new RegExp("(" + k + ")").exec(fmt);
        if (ret) {
            fmt = fmt.replace(ret[1], (ret[1].length == 1) ? (opt[k]) : (opt[k].padStart(ret[1].length, "0")))
        };
    };
    return fmt;
}

/**
 * 调用vscode原生api
 * @param data 可以是类似 {cmd: 'xxx', param1: 'xxx'}，也可以直接是 cmd 字符串
 * @param cb 可选的回调函数
 */
function callVscode(data, cb) {
    if (typeof data === 'string') {
        data = { cmd: data };
    }
    if (cb) {
        const cbid = Date.now() + '' + Math.round(Math.random() * 100000);
        callbacks[cbid] = cb;
        data.cbid = cbid;
    }
    vscode.postMessage(data);
}

window.addEventListener('message', event => {
    const message = event.data;
    switch (message.cmd) {
        case 'vscodeCallback':
            (callbacks[message.cbid] || function () { })(message.data);
            delete callbacks[message.cbid];
            break;
        default: break;
    }
});

function paddingZero(value){
    if(value<10){
        return '0'+value
    }
    return value
}
new Vue({
    el: '#app',
    data: {
        dateForShow: moment(),
        // currentDate: moment().format('YYYY-MM-DD')
    },
    mounted() {
        var day = dateFormat('YYYYmmdd',new Date())
        callVscode({cmd: 'getRecord', day: day}, result => {
            this.updateChart(result, day)
        });
        $("#datepicker").daterangepicker({
            singleDatePicker: true,
            showDropdowns: true,
            opens: 'center',
          }, (start,end,label) => {
            this.dateForShow = start
          });
    },
    computed:{
        currentDate(){
            return this.dateForShow.format('YYYY-MM-DD')
        }
    },
    watch: {
        dateForShow(val){
            var day = this.dateForShow.format('YYYYMMDD')
            callVscode({cmd: 'getRecord', day: day}, result => {
                this.updateChart(result, day)
            });
        }
    },
    methods: {
        updateChart(result,startDay){
            var hours = ['00','01', '02', '03', '04', '05', '06',
                    '07', '08', '09','10','11',
                    '12', '13', '14', '15', '16', '17',
                    '18', '19', '20', '21', '22', '23'];
            var minutes = ['00', '10', '20',
                    '30', '40', '50'];

            var data = [];
            for(var i=0;i<hours.length;i++){
                for(var j=0;j<minutes.length;j++){
                    data.push([i,j,0,[]])
                }
            }
            result.forEach(r => {
                var recordTime = new Date(r.time)
                var hour = recordTime.getHours()
                var minute = Math.ceil((recordTime.getMinutes()*60+recordTime.getSeconds())/600)-1
                if(r.type==='coding' || r.type==='debug'){
                    var cell = data[hour*6+minute]
                    cell[3].push(r.type)
                    data[hour*6+minute] = [cell[0],cell[1],cell[2]+1, Array.from(new Set(cell[3]))]
                }
            })

            data = data.map(function (item) {
                return [item[0], item[1], item[2] || '-', item[3]];
            });
            var option = {
                title: {
                    show: false,
                    text: `${startDay.substr(0,4)}-${startDay.substr(4,2)}-${startDay.substr(6,2)}`,
                    left: 'center',
                    top: 10,
                },
                tooltip: {
                    position: 'top',
                    padding: 0,
                    formatter: function(params){
                        var startHour = params.value[0]
                        var startMinute = params.value[1]
                        var endHour = startMinute<5?startHour:startHour+1
                        var endMinute = startMinute<5?(startMinute+1):0
                        var tasks = params.value[3].join("<br />")
                        var contributions = params.value[2]
                        return `
                        <table border="0" style="border-collapse:collapse;">
                            <tr>
                                <th colspan="2" style="padding: 5px;">${params.marker}${paddingZero(startHour)}:${paddingZero(startMinute*10)}-${paddingZero(endHour)}:${paddingZero(endMinute*10)}</th>
                            </tr>
                            <tr>
                                <td valign="top" style="padding: 2px 5px;">Task</td>
                                <td valign="top" align="right" style="padding: 2px 5px;">${tasks}</td>
                            </tr>
                            <tr>
                                <td style="padding: 2px 5px;">Contributions</td>
                                <td align="right" style="padding: 2px 5px;">${contributions}</td>
                            </tr>
                        </table>
                        `
                    }
                },
                animation: true,
                grid: {
                    top: 'middle',
                    left: 'center',
                    width: 640,
                    height: 160,
                    containLabel: false,
                    show: true,
                    backgroundColor: '#ebedf0',
                },
                visualMap: {
                    show: false,
                    min: 0,
                    max: 15,
                    dimension: 2,
                    calculable: false,
                    orient: 'horizontal',
                    left: 'center',
                },
                xAxis: {
                    type: 'category',
                    name: 'Hour',
                    data: hours,
                    zlevel: 100,
                    splitLine: {
                        show: true,
                        interval: 0,
                        lineStyle:{
                           color: ['rgba(255,255,255,1)'],
                           width: 1,
                           type: 'solid'
                        }
                　　},
                    axisTick: {
                        show: false
                    }
                },
                yAxis: {
                    type: 'category',
                    name: 'Minute',
                    data: minutes,
                    zlevel: 100,
                    splitLine: {
                        show: true,
                        interval: 0,
                        lineStyle:{
                            color: ['rgba(255,255,255,1)'],
                            width: 1,
                            type: 'solid'
                        }
                　　},
                    axisTick: {
                        show: false
                    }
                },
                series: [{
                    name: 'Coding record',
                    type: 'heatmap',
                    data: data,
                    label: {
                        normal: {
                            show: false
                        }
                    },
                    itemStyle: {
                        emphasis: {
                            shadowBlur: 10,
                            shadowColor: 'rgba(0, 0, 0, 0.5)'
                        }
                    }
                }]
            };

            var myChart = echarts.init(this.$refs.chart, 'dark');
            myChart.setOption(option);
        },
        getTime() {
            const hour = new Date().getHours();
            if (hour <= 8) return '早上';
            else if (hour < 12) return '上午';
            else if (hour < 14) return '中午';
            else if (hour < 18) return '下午';
            return '晚上';
        }
    }
});